### ISSUE TITLE
Scavenger hits invalid object pointer

### ISSUE THREAD

**Issue by mkustermann** [dart-lang/sdk#41722]
A flaky crash has happened (from [log](https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket.appspot.com/8881610527335131936/+/steps/test_results/0/logs/ignored_flaky_test_failure_logs/0)):
```

/============================================================\
| vm/dart/splay_test/15 broke (Pass -> Crash, expected Pass) |
\============================================================/

--- Command "vm" (took 02.000750s):
set DART_CONFIGURATION=ReleaseX64 & set DART_SUPPRESS_WER=1 & set DART_CRASHPAD_HANDLER=C:\b\s\w\ir\out\ReleaseX64\crashpad_handler.exe & out\ReleaseX64\dart.exe --verify_before_gc --verify_after_gc --ignore-unrecognized-flags --packages=C:\b\s\w\ir\.packages C:\b\s\w\ir\runtime\tests\vm\dart\splay_test.dart

exit code:
1073741845

stderr:
Verifying before marking... done.
Verifying before marking... done.
Verifying before marking...Verifying before sweeping... done.
 done.
Verifying before marking... done.
Verifying before sweeping... done.
Verifying before Scavenge... done.
Verifying after Scavenge... done.
Verifying before Scavenge... done.
Verifying after Scavenge... done.
Verifying before marking... done.
Verifying before sweeping... done.
Verifying before Scavenge... done.
Verifying after Scavenge... done.
Verifying before marking... done.
Verifying before sweeping... done.
Verifying before Scavenge... done.
Verifying after Scavenge... done.
Verifying before Scavenge... done.
Verifying after Scavenge... done.
Verifying before Scavenge... done.
Verifying after Scavenge...c:\b\s\w\ir\cache\builder\sdk\runtime\vm\heap\verifier.cc: 57: error: Invalid object pointer encountered 0x200f45bfb88

version=2.9.0-edge.afd43c430212e81a748460a4bb07c5dd6381ab55 (be) (Thu Apr 30 07:52:24 2020 +0000) on "windows_x64"
pid=5868, thread=11024, isolate=(null)(0000000000000000)
isolate_instructions=0, vm_instructions=7ff766c71010
  pc 0x00007ff766e71efc fp 0x000000740e3fec50 Dart_CurrentIsolate+0x11dcfc
-- End of DumpStackTrace
[exit     : sp(0) fp(0x740e3ff348) pc(0)]
[dart     : sp(0x740e3ff358) fp(0x740e3ff380) pc(0x200ecb21dd9) dart:core__OneByteString@0150898__allocate@0150898 ]
[dart     : sp(0x740e3ff390) fp(0x740e3ff3a8) pc(0x200ecb3f078) *dart:core__OneByteString@0150898__concatAll@0150898 ]
[dart     : sp(0x740e3ff3b8) fp(0x740e3ff3e0) pc(0x200ecb3f2a4) *dart:core__StringBase@0150898__interpolate@0150898 ]
[dart     : sp(0x740e3ff3f0) fp(0x740e3ff420) pc(0x200ecb3f4b2) *file:///C:/b/s/w/ir/runtime/tests/vm/dart/splay_test.dart_Payload_generate ]
[dart     : sp(0x740e3ff430) fp(0x740e3ff468) pc(0x200ecb3f6b4) *file:///C:/b/s/w/ir/runtime/tests/vm/dart/splay_test.dart_Payload_generate ]
[dart     : sp(0x740e3ff478) fp(0x740e3ff4b0) pc(0x200ecb3f6b4) *file:///C:/b/s/w/ir/runtime/tests/vm/dart/splay_test.dart_Payload_generate ]
[dart     : sp(0x740e3ff4c0) fp(0x740e3ff4f8) pc(0x200ecb3f69c) *file:///C:/b/s/w/ir/runtime/tests/vm/dart/splay_test.dart_Payload_generate ]
[dart     : sp(0x740e3ff508) fp(0x740e3ff540) pc(0x200ecb3f69c) *file:///C:/b/s/w/ir/runtime/tests/vm/dart/splay_test.dart_Payload_generate ]
[dart     : sp(0x740e3ff550) fp(0x740e3ff588) pc(0x200ecb3f69c) *file:///C:/b/s/w/ir/runtime/tests/vm/dart/splay_test.dart_Payload_generate ]
[dart     : sp(0x740e3ff598) fp(0x740e3ff5d0) pc(0x200ecb3a47f) file:///C:/b/s/w/ir/runtime/tests/vm/dart/splay_test.dart_Splay_insertNewNode ]
[dart     : sp(0x740e3ff5e0) fp(0x740e3ff600) pc(0x200ecb3a1ea) file:///C:/b/s/w/ir/runtime/tests/vm/dart/splay_test.dart_Splay_mysetup ]
[dart     : sp(0x740e3ff610) fp(0x740e3ff628) pc(0x200ecb3a023) file:///C:/b/s/w/ir/runtime/tests/vm/dart/splay_test.dart_Splay_main ]
[dart     : sp(0x740e3ff638) fp(0x740e3ff650) pc(0x200ecb39f63) file:///C:/b/s/w/ir/runtime/tests/vm/dart/splay_test.dart_::_main ]
[dart     : sp(0x740e3ff660) fp(0x740e3ff690) pc(0x200ecb39eb7) file:///C:/b/s/w/ir/runtime/tests/vm/dart/splay_test.dart_::_main_main ]
[dart     : sp(0x740e3ff6a0) fp(0x740e3ff6c8) pc(0x200ecb39cf7) dart:core__Closure@0150898_call ]
[dart     : sp(0x740e3ff6d8) fp(0x740e3ff710) pc(0x200ecb39007) dart:isolate_::__startIsolate@1026248_<anonymous closure> ]
[dart     : sp(0x740e3ff720) fp(0x740e3ff750) pc(0x200ecb38d2a) dart:core__Closure@0150898_call ]
[dart     : sp(0x740e3ff760) fp(0x740e3ff788) pc(0x200ecb38c3e) dart:isolate__RawReceivePortImpl@1026248__handleMessage@1026248 ]
[entry    : sp(0x740e3ff798) fp(0x740e3ff8a8) pc(0x200ebf01315)]

--- Re-run this test:
python tools/test.py -n dartk-win-release-x64 vm/dart/splay_test/15
```

Tentatively assigning to @rmacnak-google . Do you think this could be related to the paged new space?

**Comment by rmacnak-google** [comment]
The version in the crash (afd43c430212e81a748460a4bb07c5dd6381ab55) does not have paged new space. The bad pointer has new-space alignment, so my guess would be a missing store barrier.

**Comment by rmacnak-google** [comment]
Probably fixed by 3844e74adbcc55a933180aef9bb4408c3b851cfa or 32dfefbea17d92302147bec203324606200315f9
